### DISCUSSION TITLE
Shallow Water Equations and Moose Vector Variables

### DISCUSSION THREAD

**Discussion by DamynChipman** [idaholab/moose#19356] [Q&A Modules: General] [answered]
I am attempting to model the shallow water equations (SWE) with Moose (I'm a PhD student at Boise State University). I've used Moose for a variety of other geophysics, but the SWE are slightly more complicated. I mention that to say that I'm familiar with Moose, having gone through most of the tutorials.

Here's my problem setup. I have the strong form that I use to generate the weak form of the equations. The weak form takes the following form (using inner product notation):

`(dqdt, \psi) + <F(q) * n, \psi> - (F(q), \nabla \psi) - (b + s, \psi) = 0`

`\psi` is the test function.
`q` is the variable vector: `[h, hu, hv]`.
`F(q)` is the flux tensor: `[ [hu, hv], [hu^2 + (1/2)gh^2, huv], [huv, hv^2 + (1/2)gh^2] ]`.
`h` is water height.
`u` and `v` are velocities.
`b` and `s` are bathymetry and (friction) source terms.

These would be my questions:

How do I work with vector valued variables in Moose? What classes can I work with/ derive from?
What about for vectors like `q` where the entries aren't necessarily physically related (like components of a velocity field)?
Would I be able to use some of the tensor functionality to do this? I have briefly looked over it.

I will continue to pour over tutorials and documentation, but any guidance would be appreciated.

And finally, I read somewhere that a SWE module is in the works. Any word on progress there?

Many thanks!

**Comment by GiudGiud** [comment, marked as answer]
Hello

~~First, since you only have 3 components and the equation seems to be writable in a vector-formulation, vector is the right choice and should give you better performance over implementing 3 scalar equations.~~
I m a little concerned that having h and hu hv, which could be of different orders of magnitude since they have a different unit / physical meaning, will lead to convergence/scaling issues. You won't be able to scale the equations separately if you lump them in a vector
 
I m not sure if we have documentation specially for implementing vector equations. If you find it please let me know. The process is mostly the same, you create a vector variable, then vector kernels to implement your equation etc. There is a `VectorKernel` base class you can inherit from for your kernels. 
If you have a tensor anisotropic diffusivity for example, you will be able to use the tensor utilities (in utils/) for that. 

We support the lagrange variables as vector variables. There are vector versions of kernels, bcs, auxkernels etc. 
An example of a vector BC:
https://mooseframework.inl.gov/source/bcs/VectorDirichletBC.html

vector kernel:
https://mooseframework.inl.gov/moose/source/kernels/VectorDiffusion.html

AuxKernel has an example of a vector aux in the text:
https://mooseframework.inl.gov/syntax/AuxKernels/

Where did you read about the SWE? It d be easier to tag the person who was talking about it here

Guillaume

**Reply by lindsayad** [reply]
Yes I would only ever use a vector variable if the governing equation can logically be written in vector form like the momentum equation in Navier-Stokes. Speaking of which, the AD version of finite element incompressible Navier-Stokes in the `navier_stokes` module is probably one of the better places to look if you're curious about using vector variables.

**Reply by lindsayad** [reply]
Our shallow water equation modeling with finite elements never really got off the ground. For modeling of hyperbolic-character equations we're going more and more towards our finite volume implementation.

**Reply by DamynChipman** [reply]
Thanks for the resources on the vector variables and kernels. I'll dive into those objects to see what I can get done. All of the tutorials work with scalar variables and kernels, so thanks for pointing me towards the vector stuff. I hadn't considered the different orders of magnitude in `h` and `u`/`v`; I will keep that in mind moving forward.

I saw in one of the tutorial presentations that SWE was a work in progress. I don't know who was working on it (or if that is still the case).

I did see that Moose supports FV modeling, but I haven't looked into it much. I'll have to do that. That's how we normally solve the SWE anyways.

Regardless, thanks for the direction! If I get any good results/progress, or if I have any further questions, I'll post them here.

**Reply by dschwen** [reply]
Yeah, I implemented a shallow water model years ago, but I didn't know what I was doing and It had horrible artifacts with kilometer waves after some time :-D. I suppose the formulation was not stable with regular Galerkin FEM. 🤷‍♂️

**Reply by dschwen** [reply]
Here is the tidal model (it was done just for fun!)

https://user-images.githubusercontent.com/202302/141403142-772cdf1a-36da-412b-aed2-8103540efece.mp4

**Reply by lindsayad** [reply]
@aeslaughter had also worked on SWE

**Reply by dschwen** [reply]
Resurrecting an old thread here, but I'm very close to have a DG based shallow water module ready for a pull request...

**Reply by lindsayad** [reply]
<img width="300" height="168" alt="image" src="https://github.com/user-attachments/assets/0d5335f6-f736-4134-b750-67d11fd96db9" />

I'm stoked to here about this. What project is this for?

**Reply by lindsayad** [reply]
Related to shallow water equations, we have HDG Navier-Stokes now. You can see a newsletter entry about it at https://mooseframework.inl.gov/newsletter/2025/2025_06.html. I would really like a non-hybridized DG implementation of Navier-Stokes as well

**Reply by dschwen** [reply]
> I'm stoked to here about this. What project is this for?

Its... for the AI pilot program. The module is made using Codex as a trial for how far I can get running on ignorance and a few million API tokens...

And it turns out: pretty far.

I need to clean up some idiosyncrasies in the code, HLLC is just 1D right now, but HLL(E) looks  pretty good so far.

**Reply by lindsayad** [reply]
Have you looked at the HLLC implementation in FV? https://mooseframework.inl.gov/source/fvkernels/CNSFVHLLCBase.html is probably the best doc page
